include/transport/stream_id.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <atomic>
+#include <stdint.h>
+
+namespace qtransport {
+  class StreamId {
+    using type = uint64_t;
+    friend struct std::atomic<StreamId>;
+
+  public:
+    StreamId() = default;
+    ~StreamId() = default;
+
+    constexpr StreamId(type value, bool is_server, bool is_unidirectional)
+        : _is_server{is_server},
+          _is_unidirectional{is_unidirectional} {
+      adjust(value);
+    }
+
+    constexpr StreamId(type value)
+        : _value{value},
+          _is_server{value & 0b01},
+          _is_unidirectional{value & 0b10} {}
+
+    constexpr StreamId(const StreamId &) = default;
+    constexpr StreamId(StreamId &&) = default;
+
+    constexpr StreamId &operator=(const StreamId &) = default;
+    constexpr StreamId &operator=(StreamId &&) = default;
+    constexpr StreamId &operator=(const type &value) {
+      adjust(value);
+      return *this;
+    }
+    constexpr StreamId &operator=(type &&value) {
+      _value = std::move(value);
+      adjust(_value);
+      return *this;
+    }
+
+    constexpr operator type() const {
+      return _value;
+    }
+
+    constexpr StreamId &operator++() noexcept {
+      adjust(_value + 4);
+      return *this;
+    }
+
+    constexpr StreamId operator++(int) noexcept {
+      StreamId id(*this);
+      ++*this;
+      return id;
+    }
+
+  private:
+    constexpr void adjust(type value) {
+      _value = (value & (~0x0u << 2)) | _is_server | (_is_unidirectional * 2);
```

Comment:
Not showing binary literals here with ternary because:
- `_is_server` can only ever be `0b00` or `0b01`, so use the value directly
- `_is_unidirectional` can only ever be `0b00` or `0b01`, so multiply by 2 to get either `0b00` or `0b10`

Reply:
are we creating QUIC StreamIds outside QUIC ?

Reply:
Oops, that's a good point